src/porepy/models/momentum_balance.py

```diff
@@ -845,10 +846,60 @@ def bc_values_stress(self, boundary_grid: pp.BoundaryGrid) -> np.ndarray:
     def update_all_boundary_conditions(self) -> None:
         """Set values for the displacement and the stress on boundaries."""
         super().update_all_boundary_conditions()
+        self.update_boundary_condition(self.stress_keyword, self.bc_values_stress)
+
+    def update_boundary_values_primary_variables(self) -> None:
+        """Updates the displacement on the boundary, as the primary variable for
+        mechanics."""
+        super().update_boundary_values_primary_variables()
         self.update_boundary_condition(
             self.displacement_variable, self.bc_values_displacement
         )
-        self.update_boundary_condition(self.stress_keyword, self.bc_values_stress)
+
+
+class InitialConditionsMomentumBalance(pp.InitialConditionMixin):
+    """Mixin for providing initial values for displacement."""
+
+    displacement: Callable[[pp.SubdomainsOrBoundaries], pp.ad.Operator]
+    """See :class:`VariablesMomentumBalance`."""
+
+    def set_initial_values_primary_variables(self) -> None:
+        """Method to set initial values for displacement at iterate index 0.
+
+        See also:
+
+            - :meth:`initial_displacement`
+
+        """
+        # super call for compatibility with multi-physics
+        super().set_initial_values_primary_variables()
+
+        for sd in self.mdg.subdomains():
+            # Displacement is only defined on grids with ambient dimension.
+            if sd.dim == self.nd:
+                # Need to cast the return value to variable, because it is typed as
+                # operator.
+                self.equation_system.set_variable_values(
+                    self.initial_displacement(sd),
```

Comment:
Due to lack of knowledge what is considered primary in mechanics. Please let me know and I will add respective code.

Reply:
Summary of discussion: See VariablesMomentumBalance, or whatever it is called, for an overview of primary variables.

Reply:
All variables in base models (mass, energy, momentum) are now covered by the IC mixin.

For mass and energy:

While p and T are clearly primary, I took the liberty to put the initialization of various interface and well fluxes on the level of `initial_condition`, and not `set_initial_values_primary_variables`. This way the interface and well fluxes get initialized after the primary variables (p, T, u, fractions,...) I think this will be useful in use cases which compute the initial interface fluxes based on the constitutive laws/interface equations.

For momentum:

I followed your instructions and put the initialization of all, `displacement` `interface_displacement` and `contact_traction`, into the sub-routine `set_initial_values_primary_variables`. Please check this carefully, since  I touched code whose physics I do not understand. There was a default initialzation routine for contact traction in the `SolutionStrategyMomentumBalance`, which I moved to `InitialConditionsMomentumBalance`.

Note on tests:

Various model setups still put the complete initialization routine in a single overload of `initial_condition`. We should open a respective issue to have that changed.